0003-lightning-as-alpha-ledger.md

```diff
@@ -0,0 +1,158 @@
+# LND as Alpha Ledger as decribed in RFC003
+
+* Authors: Philipp Hoenisch
+* Date: 2019-04-24
+
+Tracking issue: TBA
+
+## Context
+
+The **Hold invoice feature** was introduced in the following PR: [lightningnetwork/lnd/pull/2022](https://github.com/lightningnetwork/lnd/pull/2022).
+> A hold invoice is a new type of invoice that triggers a different flow at the receiver end.
+> Instead of immediately locking in and settling the htlc when the payment arrives, the htlc for a hold invoice is only locked in and not yet settled.
+> At that point, it is not possible anymore for the sender to revoke the payment, but the receiver still can choose whether to settle or cancel the htlc and invoice.
+
+This allows us to trade Bitcoin on the Lightning Network (LN) for an asset _X_ on a different ledger where LN is the _alpha ledger_ as per the definition in [RFC003](https://github.com/comit-network/RFCs/blob/master/RFC-003-SWAP-Basic.md).
+For the other way around, Bitcoin on the Lightning Network is the _beta asset/ledger_, we can make use of `createinvoice`.
+
+### Standard _hold invoice_ process
+
+> Assumption: Alice wants to pay Bob _X_ Bitcoin using the LN but Bob does not generate the preimage - Alice does it.
+1) Alice generates a secret preimage `p` and sends the hash of it `h(p)` to Bob.
+2) Bob calls `addholdinvoice` with an amount `X` and the hashed preimage `h(p)` ([more details](https://github.com/lightningnetwork/lnd/blob/aa1cd04dbf07a9195d5ada752f383988d8d01fa7/cmd/lncli/invoicesrpc_active.go#L142)).
+3) Bob sends the generated invoice to Alice.
+4) Alice now can pay the invoice using `sendpayment`.
+5) As soon as Alice is happy (because she is in a good mode), she tells Bob the secret (or he learns it otherwise, see below).
+6) Bob settles the payment by calling `settleInvoice` ([more details](https://github.com/lightningnetwork/lnd/blob/aa1cd04dbf07a9195d5ada752f383988d8d01fa7/cmd/lncli/invoicesrpc_active.go#L53)).
```

Comment:
I would not consider the spike complete until this is tested and we know how it looks like (at least on command line).

Reply:
Fair enough, I already started playing around with it :)

Reply:
I can confirm, the following works: 


`secret=0000000000000000000000000000000000000000000000000000000000000001`

`sha256=ec4916dd28fc4c10d78e287ca5d9cc51ee1ae73cbfde08c6b37324cbfaac8bc5`

---

```bash
$bob: lncli --network=simnet addholdinvoice ec4916dd28fc4c10d78e287ca5d9cc51ee1ae73cbfde08c6b37324cbfaac8bc5 --amt 10
```



---



```bash
$alice: lncli --network=simnet payinvoice lnsb100n....
```


---

```bash
$bob: `lncli --network=simnet lookupinvoice 4146873...
{
    "memo": "",
    "receipt": null,
    "r_preimage": null,
    ...
    "amt_paid_msat": "10000",
    "state": "ACCEPTED"
}
```

---

```bash
$bob: lncli --network=simnet settleinvoice 0000000000000000000000000000000000000000000000000000000000000001
``` 

---

```bash
$bob: `lncli --network=simnet lookupinvoice ec4916dd28fc4c10d78e287ca5d9cc51ee1ae73cbfde08c6b37324cbfaac8bc5
{
    "memo": "",
    "receipt": null,
    "r_preimage": "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAE=",
    "r_hash": "7EkW3Sj8TBDXjih8pdnMUe4a5zy/3gjGs3Mky/qsi8U=",
    "value": "10",
    "settled": true,
    "creation_date": "1556272625",
    "settle_date": "1556272661",
    "payment_request": "lnsb100n1pwv9403pp5a3y3dhfgl3xpp4uw9p72tkwv28hp4eeuhl0q334nwvjvh74v30zsdqqcqzpgjemg8uy8y7ej0q6lmxvkfhwfrjnesv6jryk46m5kcfcd2drykskq97c2ad9lvp9d6mmyt0r6rhp26e2cmrqd9qgc88rf5l58hz6ntxgqc3azkn",
    "description_hash": null,
    "expiry": "3600",
    "fallback_addr": "",
    "cltv_expiry": "40",
    "route_hints": [
    ],
    "private": false,
    "add_index": "7",
    "settle_index": "1",
    "amt_paid": "10000",
    "amt_paid_sat": "10",
    "amt_paid_msat": "10000",
    "state": "SETTLED"
}
```